### PR TITLE
[SPEC] Add fileCount to dataset stats facets

### PR DESCRIPTION
**Pull request by dolfinus**
### Problem

👋 Thanks for opening a pull request! Please include a brief summary of the problem your change is trying to solve, or bug fix. If your change fixes a bug or you'd like to provide context on why you're making the change, please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) as follows:

Closes: #2550

### Solution

Please describe your change as it relates to the problem, or bug fix, as well as any dependencies. If your change requires a schema change, please describe the schema modification(s) and whether it's a _backwards-incompatible_ or _backwards-compatible_ change, then select one of the following:

> **Note:** All schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [X] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

#### One-line summary:

Adds "fileCount" field to DataQualityMetricsInputDatasetFacet and OutputStatisticsOutputDatasetFacet specification

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [X] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [X] Your changes are accompanied by tests (_if relevant_)
- [X] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [X] Your comment includes a one-liner for the changelog about the specific purpose of the change (_if necessary_)
- [X] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [ ] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2023 contributors to the OpenLineage project

### CODE REVIEW

File: proposals/168/making_spark_visitors_extensible.md [hunk 1]
```diff
@@ -259,10 +259,10 @@ public class OutputStatisticsFacetBuilder implements CustomFacetBuilder<StageInf
...
     public void accept(StageInfo event, BiConsumer<String, ? super OpenLineage.OutputStatisticsOutputDatasetFacet> consumer) {
         OpenLineage.OutputStatisticsOutputDatasetFacet statsFacet =
-                openLineage.newOutputStatisticsOutputDatasetFacet(
-                        event.taskMetrics().outputMetrics().recordsWritten(),
-                        event.taskMetrics().outputMetrics().bytesWritten()
-                );
+                openLineage.newOutputStatisticsOutputDatasetFacetBuilder()
+                  .rowCount(event.taskMetrics().outputMetrics().recordsWritten())
+                  .size(event.taskMetrics().outputMetrics().bytesWritten())
+                  .build();
```
Comment (JDarDagran) [inline proposals/168/making_spark_visitors_extensible.md:265]: I don't think we need to update proposals :)

File: spec/facets/DataQualityMetricsInputDatasetFacet.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "$schema": "https://json-schema.org/draft/2020-12/schema",
-  "$id": "https://openlineage.io/spec/facets/1-0-0/DataQualityMetricsInputDatasetFacet.json",
+  "$id": "https://openlineage.io/spec/facets/1-1-0/DataQualityMetricsInputDatasetFacet.json",
```
Comment (JDarDagran) [inline spec/facets/DataQualityMetricsInputDatasetFacet.json:3]: I'd rather make it 1-0-1 bump according to https://github.com/OpenLineage/OpenLineage/blob/main/spec/Versioning.md#L29

File: spec/facets/OutputStatisticsOutputDatasetFacet.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "$schema": "https://json-schema.org/draft/2020-12/schema",
-  "$id": "https://openlineage.io/spec/facets/1-0-0/OutputStatisticsOutputDatasetFacet.json",
+  "$id": "https://openlineage.io/spec/facets/1-1-0/OutputStatisticsOutputDatasetFacet.json",
```
Comment (JDarDagran) [inline spec/facets/OutputStatisticsOutputDatasetFacet.json:3]: https://github.com/OpenLineage/OpenLineage/blob/main/spec/Versioning.md#L29

**Review by JDarDagran — CHANGES_REQUESTED** [review]

File: proposals/168/making_spark_visitors_extensible.md (hunk 1 above)
Comment (dolfinus) [reply proposals/168/making_spark_visitors_extensible.md]: Oops, sorry)

**Comment by dolfinus** [comment]
Adding third field to OutputStatisticsOutputDatasetFacet lead to replacing method `newOutputStatisticsOutputDatasetFacetBuilder` with `newOutputStatisticsOutputDatasetFacet`. This could introduce compatibility issues for users which use Java client.

**Comment by JDarDagran** [comment]
> Changing number of fields in OutputStatisticsOutputDatasetFacet from 2 to 3 lead to replacing method `.newOutputStatisticsOutputDatasetFacet(rowCount, size)` with `.newOutputStatisticsOutputDatasetFacetBuilder().rowCount(...).size(...)`. This could introduce compatibility issues for users which use Java client.

I don't understand why? I just checked and 
```java
public OutputStatisticsOutputDatasetFacet newOutputStatisticsOutputDatasetFacet(Long rowCount,
      Long size, Long fileCount) {
    return new OutputStatisticsOutputDatasetFacet(this.producer, rowCount, size, fileCount);
  }
```
gets generated as well.

**Comment by dolfinus** [comment]
Ah, I see, now method accepts 3 arguments instead of 2. Without default value for filesCount, users still have to update their code

**Comment by mobuchowski** [comment]
Having exact number of arguments is by design. If you don't care you can use builder as solution here.

**Comment by JDarDagran** [comment]
LGTM :+1:
@dolfinus, may I kindly ask how did you find the need to add this field?
Would it be worth/possible to add the statistic somewhere in any of the existing integration?

**Review by mobuchowski — APPROVED** [review]

**Comment by mobuchowski** [comment]
Thanks for contribution @dolfinus .

**Pull request merged**